RFC.md

```diff
@@ -1,12 +1,234 @@
+
+- This has also been implemented as part of a GSoC effort to unify
+Chapel and Polly. 
+
+- Molly, the distributed version of Polly written by Michael Kruse for his PhD
+  also implemented a similar scheme. In his use-case, guessing as we perform
+  optimistically was not possible, so this kind of intrinsic was _necessary_
+  for the application, not just _good to have_. More details are available
+  in his PhD thesis: [Lattice QCD Optimization and Polytopic Representations of Distributed Memory](https://tel.archives-ouvertes.fr/tel-01078440).
+  In particular, Chapter 9 contains a detailed discussion.
+
+# Representations
+
+## Instruction / Intrinsic
+
+### Syntax
+```
+<result> = multidim_array_index <ty> <ty>* <ptrval> {<stride>, <idx>}*
+```
+
+### Overview:
+The `multidim_array_index` instruction is used to get the address of 
+an element from an array. It performs address calcuation only and 
+does not access memory. It is similar to `getelementptr`. However, it imposes
+additional semantics which allows the optimiser to provide better optimisations
+than `getlementptr`.
+
+
+
+### Arguments:
+The first argument is always a type used as the basis for the calculations. The
+second argument is always a pointer, and is the base address to start the
+calculation from. The remaining arguments are a list of pairs. Each pair
+contains a dimension stride, and an offset with respect to that stride.
+
+
+### Semantics:
+
+`multidim_array_index` represents a multi-dimensional array index, In particular, this will
+mean that we will assume that all indeces `<idx_n>` are non-negative.
+
+
+##### Address computation:
+Consider an invocation of `multidim_array_index`:
+
+```
+<result> = multidim_array_index <ty> <ty>* <ptrval> <str_0>, <idx_0>, <str_1> <idx_1>, ..., <str_n> <idx_n>
+```
+
+If the pairs are denoted by `(str_i, idx_i)`, where `str_i` denotes the stride
+and `idx_i` denotes the index of the ith pair, then the final address (in bytes)
+is computed as:
+
+```
+addressof(ptrval) + len(ty) * [(str_0 * idx_0) + (str_1 * idx_1) + ... (str_n * idx_n)]
+```
+
+## Transitioning to `multidim_array_index`: Allow `multidim_array_index` to refer to a GEP instruction:
+
+```llvm
+i1 foo(... arr):
+    %gep = getelementptr ...
+    %multidim = multidim_array_array(%size0, %size1, %ix0, %ix1, %arr, %gep)
+    %arr.load = load %multidim
+```
+
+We will ensure that calls such as `isGEP` and `dyn_cast<GEP>` will proxy
+through `%gep` to return the correct values. This will ensure that we don't
+lose the current optimiser when trying to teach the optimiser about
+`multidim_array_index`.
+
+
+# Caveats
+
+Currently, we assume that the array shape is immutable. However, we will need to deal with 
+being able to express `reshape` like primitives where the array shape can be mutated. However,
+this appears to make expressing this information quite difficult: We now need to attach the shape
+information to an array per "shape-live-range". 
```

Comment:
I'm happy to include the caveats here but I still think it'd be useful to point out that the representation allows it (by say loading the stride value into a register and then using it in multidim_array_index). In other words, I think the caveat here is that we havn't *writing optimizations* that allow the array shape to change (rather than that the representation proposed can't represent it).